core/src/test/java/io/grpc/transport/MessageFramerTest.java

```diff
@@ -149,60 +145,134 @@ public void flushIsIdempotent() {
     verify(sink).deliverFrame(frameCaptor.capture(), eq(false));
-    List<Byte> buffer = frameCaptor.getValue();
+    ByteWritableBuffer buffer = frameCaptor.getValue();
     assertEquals(1005, buffer.size());
-    assertEquals(Bytes.asList(new byte[] {0, 0, 0, 3, (byte) 232}), buffer.subList(0, 5));
+
+    byte data[] = new byte[1005];
+    data[3] = 3;
+    data[4] = (byte) 232;
+
+    assertEquals(toWriteBuffer(data, transportFrameSize), buffer);
     verifyNoMoreInteractions(sink);
   }
 
   @Test
   public void compressed() throws Exception {
     final int transportFrameSize = 100;
-    MessageFramer framer = new MessageFramer(copyingSink, transportFrameSize,
-        MessageFramer.Compression.GZIP);
+    MessageFramer framer =
+            new MessageFramer(copyingSink, allocator, transportFrameSize, Compression.GZIP);
     writePayload(framer, new byte[1000]);
     framer.flush();
     verify(sink).deliverFrame(frameCaptor.capture(), eq(false));
-    List<Byte> buffer = frameCaptor.getValue();
+    ByteWritableBuffer buffer = frameCaptor.getValue();
     // It should have compressed very well.
     assertTrue(buffer.size() < 100);
     // We purposefully don't check the last byte of length, since that depends on how exactly it
     // compressed.
-    assertEquals(Bytes.asList(new byte[] {1, 0, 0, 0}), buffer.subList(0, 4));
-    verifyNoMoreInteractions(sink);
+    assertEquals(1, buffer.data[0]);
+    assertEquals(0, buffer.data[1]);
+    assertEquals(0, buffer.data[1]);
+    assertEquals(0, buffer.data[1]);
+  }
+
+  private static WritableBuffer toWriteBuffer(byte[] data) {
+    return toWriteBuffer(data, TRANSPORT_FRAME_SIZE);
+  }
+
+  private static WritableBuffer toWriteBuffer(byte[] data, int maxFrameSize) {
+    ByteWritableBuffer buffer = new ByteWritableBuffer(maxFrameSize);
+    buffer.write(data, 0, data.length);
+    return buffer;
   }
 
   private static void writePayload(MessageFramer framer, byte[] bytes) {
     framer.writePayload(new ByteArrayInputStream(bytes), bytes.length);
   }
 
+  private static class ByteWritableBuffer implements WritableBuffer {
+    private final byte[] data;
+    private int writeIdx;
+
+    ByteWritableBuffer(int maxFrameSize) {
+      data = new byte[maxFrameSize];
+    }
+
+    @Override
+    public void write(byte[] bytes, int srcIndex, int length) {
+      if (writeIdx + length > data.length) {
+        // This should never happen if the framer is implemented correctly.
+        throw new RuntimeException("maxFrameSize exceeded");
+      }
+      for (int i = 0; i < length; i++) {
+        data[writeIdx++] = bytes[srcIndex+i];
+      }
+    }
+
+    @Override
+    public int writableBytes() {
+      return data.length - writeIdx;
+    }
+
+    @Override
+    public int readableBytes() {
+      return writeIdx;
+    }
+
+    @Override
+    public void release() {
+    }
+
+    int size() {
+      return writeIdx;
+    }
+
+    @Override
+    public boolean equals(Object buffer) {
```

Comment:
I know this is only used in the tests, but not having hashCode() is asking for trouble.

It would also not be unreasonable to fully implement equals with the addition of "if (!(buffer instanceof ByteWritableBuffer)) return false"